### ISSUE TITLE
[Bug]: Contradictory results on metrics evaluation using monitor and log

### ISSUE THREAD

**Issue by fede72bari** [DLR-RM/stable-baselines3#1465]
### 🐛 Bug

I am testing PPO on CartPole. I increased the max steps of CartPole to 500000 just to see how much PPO can learn. And I traced the episode rewards with the monitor wrapper.

Running over 300000 steps the average reward/length got stabilized around 2160 ... but is it really true?

```
-----------------------------------------
| rollout/                |             |
|    ep_len_mean          | 2.16e+03    |
|    ep_rew_mean          | 2.16e+03    |
| time/                   |             |
|    fps                  | 545         |
|    iterations           | 147         |
|    time_elapsed         | 552         |
|    total_timesteps      | 301056      |
| train/                  |             |
|    approx_kl            | 0.008984157 |
|    clip_fraction        | 0.108       |
|    clip_range           | 0.2         |
|    entropy_loss         | -0.418      |
|    explained_variance   | -0.177      |
|    learning_rate        | 0.0003      |
|    loss                 | -0.0101     |
|    n_updates            | 1460        |
|    policy_gradient_loss | -0.00287    |
|    value_loss           | 0.00575     |
-----------------------------------------
```

When I show the monitor graph I get this

![image](https://user-images.githubusercontent.com/46624075/233865877-6bad48bc-90c5-4c60-90a6-3d6039b20d0c.png)

the rewards appear almost zero, but a couple of outliners. How much zero? Could be around 2k? I showed the results in a Plotly graph after cleaning the CSV file and erasing the first remarked line that prevents pandas from correctly importing it:

```
with open(log_dir + "monitor.csv", 'r') as fin:
    data = fin.read().splitlines(True)
with open(log_dir + "monitor.csv", 'w') as fout:
    fout.writelines(data[1:])
df = pd.read_csv(log_dir + "monitor.csv")
pd.options.plotting.backend = "plotly"
fig = df['r'].plot.line()
fig.show()
```

this is the result

![image](https://user-images.githubusercontent.com/46624075/233866108-019817b1-8425-4453-b610-13abb2d234a4.png)

where the flat line close to zero is mainly between 10 and 80 reward/steps per episode; very far from the >2k ones declared by the log. Besides,

```
df = pd.read_csv(log_dir + "monitor.csv")
df.describe()
```

returns

```
	r	l	t
count	247.000000	247.000000	247.000000
mean	893.886640	893.886640	14.752001
std	8923.185641	8923.185641	46.951439
min	10.000000	10.000000	0.033821
25%	18.000000	18.000000	1.704344
50%	34.000000	34.000000	5.251336
75%	121.500000	121.500000	12.600194
max	115433.000000	115433.000000	405.454179
```

So where is the truth? thanks.

### To Reproduce

Here the link of the netbook https://www.kaggle.com/code/federicobari/ppo-stable-baseline3

```
import numpy as np
import gym as gym
from gym import spaces
import os
import torch as th
from stable_baselines3.common.monitor import Monitor
from stable_baselines3.common.results_plotter import load_results, ts2xy
from stable_baselines3 import PPO
from stable_baselines3.common import results_plotter
import plotly 
import matplotlib.pyplot as plt
import pandas as pd


# Create log dir
log_dir = "/tmp/gym4/"
os.makedirs(log_dir, exist_ok=True)

env = gym.make("CartPole-v1")
env._max_episode_steps = 500000

total_steps = 300_000

# Logs will be saved in log_dir/monitor.csv
env = Monitor(env, log_dir)

policy_kwargs = dict(activation_fn=th.nn.ReLU,
                     net_arch=dict(pi=[256, 256], vf=[256, 256]))

model = PPO("MlpPolicy", env, verbose=1, policy_kwargs=policy_kwargs)
           # tensorboard_log="/ppo_cartpole_tensorboard/")
model.learn(total_timesteps=total_steps)

# Helper from the library
results_plotter.plot_results(
            [log_dir], 
            total_steps, 
            results_plotter.X_TIMESTEPS, 
            "CartPole"
            )



with open(log_dir + "monitor.csv", 'r') as fin:
    data = fin.read().splitlines(True)
with open(log_dir + "monitor.csv", 'w') as fout:
    fout.writelines(data[1:])


df = pd.read_csv(log_dir + "monitor.csv")
df.describe()

pd.options.plotting.backend = "plotly"
fig = df['r'].plot.line()
fig.show()

```

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
- [X] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [X] I have provided a minimal working example to reproduce the bug
- [X] I've used the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

**Comment by araffin** [comment]
Hello,
I'm not sure to follow what you want to achieve? or what do you expect?


> returns

You know that `ep_rew_mean` is averaged over the last 100 episodes?  (see documentation)

See https://github.com/DLR-RM/stable-baselines3/pull/1424 to modify this value.



Side remark: I would recommend using the RL Zoo for experiment and plotting (I think I will also deprecate the built-in plotting which is quite limited).

**Comment by fede72bari** [comment]
I have some contradictory results and don't know which is the correct one:
1. the SB3 log reports an average number of steps equal to 2160, while calculating the average on the monitor CSV file we get 893. Not exactly the same, isn't it? Could be a monitor bug?
2. Secondly, I just wonder if getting for most of the episodes over 300k steps just 10-280 max steps is a plausible (poor) result from training CartPole with PPO. Because again it could be a log/monitor bug considering the two ways of reporting give different results.

**Comment by araffin** [comment]
> Being averaged over 100 episodes, how could be it equal to 2160 if most of the episodes have a number of steps around 10 and 280?

because the mean is sensitive to outliers (https://araffin.github.io/post/rliable/)
```python
import numpy as np
# 100 values
n = 100
values = np.ones(n) * 10
# outlier
values[33] = 200_000
# All values are equal to 10 but one
# can you guess the mean? closer to 10 or 1000?
print(np.mean(values))
```

> 2.

It is poor (you can check your result against the RL Zoo one), and should use the deterministic policy for evaluation.
Reaching >500 in 100k training steps:

https://huggingface.co/sb3/ppo-CartPole-v1

**Comment by fede72bari** [comment]
Thank you very much, yes it might be that the 2160 average starts after the first very lucky outliner. Thank you also for mentioning RL Zoo which I didn't know about, I will read references on it. It seems very useful.
